notion-server/src/ethereum.rs

```diff
@@ -1,4 +1,77 @@
+use base64;
 use std::collections::HashMap;
+use std::convert::TryFrom;
+use snafu::{Snafu, ResultExt, Backtrace};
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    Decode {
+        backtrace: Backtrace,
+        source: base64::DecodeError,
+    },
+    OutOfBounds {
+        message: String,
+    }
+}
+
+pub mod interface_args {
+    #[derive(Debug, Default)]
+    pub struct GeneralSimulationState {
+        pub num_execution_environments: u32,
```

Comment:
Since we're the only downstream crate of this library, this feedback isn't _super_ necessary here. I'm still going to take the opportunity to write up my thoughts and maybe provide some useful Rust information for the future!

Adding a new public field to a struct is a breaking change, unless that struct already has at least one private field. This follows from the initialization syntax having to specify every field in the struct.

For public types, generally I'd suggest using one of several alternatives:

### Constructor & Everything Private

Best for structs that have few members, or members that can easily be initialized to default values. The drawbacks are long function signatures, and having to define getters for everything.

```rust
pub struct Foo {
    bar: i32,
}

impl Foo {
    pub fn new(bar: i32) -> Self {
        Self { bar }
    }

    pub fn bar(&self) -> i32 {
        self.bar
    }
}

fn main() {
    let foo = Foo::new(5);
}
```

### One Private Field

Basically the same as the previous option, except that all fields are public, except for one zero-sized item. Avoids getters and setters, but either requires a long constructor or fields which can cheaply be default initialized.

```rust
struct Foo {
    _p: (),
    pub bar: i32,
}

impl Default for Foo {
    pub fn default() -> Self {
        Foo {
            _p: (),
            bar: 0,
        }
}

fn main() {
    let foo = Foo::default();
    foo.bar = 5;
}
```

### Typed Builder

The `typed-builder` crate is my personal favourite way of initializing structs, and probably the easiest to avoid breaking changes. `typed-builder` will derive a builder type for your struct that handles construction elegantly.

```rust
#[macro_use]
extern crate typed_builder;

#[derive(TypedBuilder)]
struct Foo {
    bar: i32,
}

fn main() {
    let foo = Foo::builder()
        .bar(5)
        .build();
}
```

Reply:
Very informative comment, thanks!  Worth mentioning that as a return value it's not necessarily a breaking change to add additional public fields, but I agree that for any incoming request (say, to the web server REST api), we'd need to create the appropriate structs based eg. on the incoming JSON data.  That said, I think the "breaking" part would depend on which interface we expect folks to use.  If we want to make this `EthereumSimulation` interface usable to all, then we'd want to have the defaults, etc handled there.  However, if the boundary is the web server REST API (or other external-facing API), then it wouldn't be a breaking change to clients of that API unless our REST server couldn't or didn't provide defaults / fallbacks to provide backward compatibility if things change in the future.  I guess my point is that everything you said is correct, but we'll always need to address the possibility of breaking changes regardless, and would need to do so wherever we want to consider the start of the public-facing interface.

That said, I like the TypedBuilder pattern, seems pretty slick, and I'm certainly on board with incorporating that into our code.  And my impression from your comment is that you think we should make `EthereumSimulation` as breaking-change-proof / backward compatible as possible.  Assuming I'm interpreting that right, I also agree with you there, though for now I think it's OK to not worry about it *too* much since as you said our web server will be the only downstream consumer, and it may save time during the prototyping phase to not worry about being breaking-change compatible right off the bat.

TL;DR: I like the TypedBuilder pattern, and think it's a great suggestion to think about using it.

Reply:
> Worth mentioning that as a return value it's not necessarily a breaking change to add additional public fields.

Without a private member, any downstream crate could create an instance. So even if it wasn't intentional for some binary to be able to create a `GeneralSimulationState`, it would still be possible, and therefore be a breaking change to add a public field. Since we control the one downstream crate, it isn't an issue here.

> And my impression from your comment is that you think we should make EthereumSimulation as breaking-change-proof / backward compatible as possible.

That would be the correct impression! I am a zealot for following SemVer.